enterprise/backend/src/metabase_enterprise/semantic_search/index.clj

```diff
@@ -373,6 +385,77 @@
                               [(:id doc) (doc->t2-model doc)]))]
     (filterv (fn [doc] (some-> doc doc->t2 mi/can-read?)) docs)))
 
+(defn- filter-by-collection
```

Comment:
Thought: Could we do something to guarantee that all search implementations use the same permissions model for correctness testing?

Could we hoist this out, and apply it at the level of all results across all engines, and then treat any other filtering that may or not happen within the engine as an optimisation that eliminates early tuples?

Reply:
Not expecting in this PR! Just a note,

Ideally you just have one function to start that covers all engines to avoid implementations drifting and applying different filtering policies, and to have one function to test. 

OTOH the performance impact might mean having engine level specialisation is worth it, or may be impractical for a whole bunch of reasons (e.g. you might need a streaming result where the caller can `.close` it or something, or you might not get enough results, and we dont have that).

Reply:
Potential performance idea:

As I read it this function needs to apply a check `location` field of each collection in the input `docs`. 

Sometimes this check is dependent on attributes of the root collection (whether it is a personal collection or not).

A faster implementation over large `docs` might be to 

- `SELECT` the `location` for all collection records referenced in `docs`
- for each of those, pull out the `root` id from the location string
- `SELECT` the `personal_owner_id` for the root ids
- filter the docs using this information according to the policy

In some cases not all these steps are needed, but these steps I think satisfy the data dependencies of any filter here in `2` `SELECT` queries, rather than doing stuff per document.